Action fails on ubuntu-24.04 with "This environment is externally managed"

**Description**

The [Ubuntu 24.04 image is now in public beta](https://github.blog/changelog/2024-05-14-github-hosted-runners-public-beta-of-ubuntu-24-04-is-now-available/). Otherwise unmodified workflows that worked on `ubuntu-22.04` now fail on `ubuntu-24.04` with `error: externally-managed-environment`.

Minimal workflow to reproduce:

```yaml
name: Test

on:
  push:

jobs:
  build:
    runs-on: ubuntu-24.04
    permissions:
      contents: read
      id-token: write

    steps:
      - name: Create artefact
        run: |
            mkdir dist
            touch dist/file.txt

      - name: Sign the artifacts with Sigstore
        uses: sigstore/gh-action-sigstore-python@v2.1.1
        with:
          inputs: dist/*
```

Here's the full log of the workflow run:

```
2024-07-03T12:38:43.8015044Z Current runner version: '2.317.0'
2024-07-03T12:38:43.8051724Z ##[group]Operating System
2024-07-03T12:38:43.8052639Z Ubuntu
2024-07-03T12:38:43.8053222Z 24.04
2024-07-03T12:38:43.8053866Z LTS
2024-07-03T12:38:43.8054409Z ##[endgroup]
2024-07-03T12:38:43.8055097Z ##[group]Runner Image
2024-07-03T12:38:43.8056005Z Image: ubuntu-24.04
2024-07-03T12:38:43.8056667Z Version: 20240630.1.0
2024-07-03T12:38:43.8058336Z Included Software: https://github.com/actions/runner-images/blob/ubuntu24/20240630.1/images/ubuntu/Ubuntu2404-Readme.md
2024-07-03T12:38:43.8061120Z Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu24%2F20240630.1
2024-07-03T12:38:43.8062596Z ##[endgroup]
2024-07-03T12:38:43.8063349Z ##[group]Runner Image Provisioner
2024-07-03T12:38:43.8064352Z 2.0.370.1
2024-07-03T12:38:43.8064942Z ##[endgroup]
2024-07-03T12:38:43.8066420Z ##[group]GITHUB_TOKEN Permissions
2024-07-03T12:38:43.8068658Z Contents: read
2024-07-03T12:38:43.8069368Z Metadata: read
2024-07-03T12:38:43.8070295Z ##[endgroup]
2024-07-03T12:38:43.8074915Z Secret source: Actions
2024-07-03T12:38:43.8076022Z Prepare workflow directory
2024-07-03T12:38:43.8744996Z Prepare all required actions
2024-07-03T12:38:43.8904119Z Getting action download info
2024-07-03T12:38:44.0087330Z Download action repository 'sigstore/gh-action-sigstore-python@v2.1.1' (SHA:61f6a500bbfdd9a2a339cf033e5421951fbc1cd2)
2024-07-03T12:38:44.4755944Z Getting action download info
2024-07-03T12:38:44.5526751Z Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
2024-07-03T12:38:44.6780861Z Download action repository 'softprops/action-gh-release@v1' (SHA:de2c0eb89ae2a093876385947365aca7b0e5f844)
2024-07-03T12:38:44.9151804Z Complete job name: build
2024-07-03T12:38:44.9952532Z ##[group]Run mkdir dist
2024-07-03T12:38:44.9952911Z [36;1mmkdir dist[0m
2024-07-03T12:38:44.9953197Z [36;1mtouch dist/file.txt[0m
2024-07-03T12:38:45.0008817Z shell: /usr/bin/bash -e {0}
2024-07-03T12:38:45.0009152Z ##[endgroup]
2024-07-03T12:38:45.0486401Z ##[group]Run sigstore/gh-action-sigstore-python@v2.1.1
2024-07-03T12:38:45.0487052Z with:
2024-07-03T12:38:45.0487361Z   inputs: dist/*
2024-07-03T12:38:45.0487701Z   staging: false
2024-07-03T12:38:45.0488077Z   verify: false
2024-07-03T12:38:45.0488432Z   upload-signing-artifacts: false
2024-07-03T12:38:45.0488852Z   release-signing-artifacts: false
2024-07-03T12:38:45.0489320Z   internal-be-careful-debug: false
2024-07-03T12:38:45.0489704Z ##[endgroup]
2024-07-03T12:38:45.0705474Z ##[group]Run # NOTE: Sourced, not executed as a script.
2024-07-03T12:38:45.0706080Z [36;1m# NOTE: Sourced, not executed as a script.[0m
2024-07-03T12:38:45.0706657Z [36;1msource "${GITHUB_ACTION_PATH}/setup/setup.bash"[0m
2024-07-03T12:38:45.0738703Z shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
2024-07-03T12:38:45.0739206Z env:
2024-07-03T12:38:45.0739695Z   GHA_SIGSTORE_PYTHON_INTERNAL_BE_CAREFUL_DEBUG: false
2024-07-03T12:38:45.0740175Z ##[endgroup]
2024-07-03T12:38:47.5962677Z error: externally-managed-environment
2024-07-03T12:38:47.5963266Z 
2024-07-03T12:38:47.5963782Z × This environment is externally managed
2024-07-03T12:38:47.5964696Z ╰─> To install Python packages system-wide, try apt install
2024-07-03T12:38:47.5965591Z     python3-xyz, where xyz is the package you are trying to
2024-07-03T12:38:47.5966202Z     install.
2024-07-03T12:38:47.5966542Z     
2024-07-03T12:38:47.5967116Z     If you wish to install a non-Debian-packaged Python package,
2024-07-03T12:38:47.5968085Z     create a virtual environment using python3 -m venv path/to/venv.
2024-07-03T12:38:47.5969058Z     Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
2024-07-03T12:38:47.5969908Z     sure you have python3-full installed.
2024-07-03T12:38:47.5970397Z     
2024-07-03T12:38:47.5971245Z     If you wish to install a non-Debian packaged Python application,
2024-07-03T12:38:47.5972109Z     it may be easiest to use pipx install xyz, which will manage a
2024-07-03T12:38:47.5972954Z     virtual environment for you. Make sure you have pipx installed.
2024-07-03T12:38:47.5974016Z     
2024-07-03T12:38:47.5974506Z     See /usr/share/doc/python3.12/README.venv for more information.
2024-07-03T12:38:47.5975046Z 
2024-07-03T12:38:47.5976672Z note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
2024-07-03T12:38:47.5978420Z hint: See PEP 668 for the detailed specification.
2024-07-03T12:38:47.6322452Z ##[error]Process completed with exit code 1.
2024-07-03T12:38:47.6511283Z Cleaning up orphan processes
```

Using `actions/setup-python@v5` before `sigstore/gh-action-sigstore-python@v2.1.1` makes the problem go away and the workflow completes successfully.

**Version**

`sigstore/gh-action-sigstore-python@v2.1.1`

--- a/.github/workflows/selftest.yml
+++ b/.github/workflows/selftest.yml
@@ -19,6 +19,8 @@ jobs:
           - ubuntu-latest
           - macos-latest
           - windows-latest
+          # TODO: Can be removed when 24.04 becomes ubuntu-latest.
+          - ubuntu-24.04
     runs-on: ${{ matrix.os }}
     if: (github.event_name != 'pull_request') || !github.event.pull_request.head.repo.fork
     steps:
@@ -27,6 +29,28 @@ jobs:
         if: ${{ matrix.os != 'ubuntu-latest' }}
         with:
           python-version: "3.x"
+      - name: Sign artifact and publish signature
+        uses: ./
+        id: sigstore-python
+        with:
+          inputs: ./test/artifact.txt
+          internal-be-careful-debug: true
+      - name: Check outputs
+        shell: bash
+        run: |
+          [[ -f ./test/artifact.txt.sigstore.json ]] || exit 1
+
+  selftest-runner-python:
+    strategy:
+      matrix:
+        os:
+          - ubuntu-latest
+          # TODO: Can be removed when 24.04 becomes ubuntu-latest.
+          - ubuntu-24.04
+    runs-on: ${{ matrix.os }}
+    if: (github.event_name != 'pull_request') || !github.event.pull_request.head.repo.fork
+    steps:
+      - uses: actions/checkout@v4
       - name: Sign artifact and publish signature
         uses: ./
         id: sigstore-python

--- a/action.py
+++ b/action.py
@@ -86,12 +86,12 @@ def _download_ref_asset(ext):
 
 
 def _sigstore_sign(global_args, sign_args):
-    return ["python", "-m", "sigstore", *global_args, "sign", *sign_args]
+    return [sys.executable, "-m", "sigstore", *global_args, "sign", *sign_args]
 
 
 def _sigstore_verify(global_args, verify_args):
     return [
-        "python",
+        sys.executable,
         "-m",
         "sigstore",
         *global_args,

--- a/action.yml
+++ b/action.yml
@@ -83,6 +83,7 @@ runs:
   using: "composite"
   steps:
     - name: Set up sigstore-python
+      id: setup
       run: |
         # NOTE: Sourced, not executed as a script.
         source "${GITHUB_ACTION_PATH}/setup/setup.bash"
@@ -93,12 +94,13 @@ runs:
     - name: Run sigstore-python
       id: sigstore-python
       run: |
-        "${GITHUB_ACTION_PATH}/.action-env/bin/python" \
+        "${VENV_PYTHON_PATH}" \
           "${GITHUB_ACTION_PATH}/action.py" \
           "${GHA_SIGSTORE_PYTHON_INPUTS}"
       env:
         # The year is 2023, and nonsense like this is still necessary on Windows.
         PYTHONUTF8: "1"
+        VENV_PYTHON_PATH: "${{ steps.setup.outputs.venv-python-path }}"
         GHA_SIGSTORE_PYTHON_IDENTITY_TOKEN: "${{ inputs.identity-token }}"
         GHA_SIGSTORE_PYTHON_SIGNATURE: "${{ inputs.signature }}"
         GHA_SIGSTORE_PYTHON_CERTIFICATE: "${{ inputs.certificate }}"

--- a/action.yml
+++ b/action.yml
@@ -93,7 +93,9 @@ runs:
     - name: Run sigstore-python
       id: sigstore-python
       run: |
-        ${GITHUB_ACTION_PATH}/action.py "${GHA_SIGSTORE_PYTHON_INPUTS}"
+        "${GITHUB_ACTION_PATH}/.action-env/bin/python" \
+          "${GITHUB_ACTION_PATH}/action.py" \
+          "${GHA_SIGSTORE_PYTHON_INPUTS}"
       env:
         # The year is 2023, and nonsense like this is still necessary on Windows.
         PYTHONUTF8: "1"

--- a/setup/setup.bash
+++ b/setup/setup.bash
@@ -45,6 +45,11 @@ min_vers=$(cut -d '.' -f2 <<< "${vers}")
 
 [[ "${maj_vers}" == "3" && "${min_vers}" -ge 7 ]] || die "Bad Python version: ${vers}"
 
-python -m pip install --requirement "${GITHUB_ACTION_PATH}/requirements.txt"
+# If the user didn't explicitly configure a Python version with
+# `actions/setup-python`, then we might be using the distribution's Python and
+# therefore be subject to PEP 668. We use a virtual environment unconditionally
+# to prevent that kind of confusion.
+python -m venv "${GITHUB_ACTION_PATH}/.action-env"
+"${GITHUB_ACTION_PATH}/.action-env/bin/python" -m pip install --requirement "${GITHUB_ACTION_PATH}/requirements.txt"
 
-debug "sigstore-python: $(python -m sigstore --version)"
+debug "sigstore-python: $("${GITHUB_ACTION_PATH}/.action-env/bin/python" -m sigstore --version)"

--- a/setup/setup.bash
+++ b/setup/setup.bash
@@ -50,6 +50,18 @@ min_vers=$(cut -d '.' -f2 <<< "${vers}")
 # therefore be subject to PEP 668. We use a virtual environment unconditionally
 # to prevent that kind of confusion.
 python -m venv "${GITHUB_ACTION_PATH}/.action-env"
-"${GITHUB_ACTION_PATH}/.action-env/bin/python" -m pip install --requirement "${GITHUB_ACTION_PATH}/requirements.txt"
 
-debug "sigstore-python: $("${GITHUB_ACTION_PATH}/.action-env/bin/python" -m sigstore --version)"
+# Annoying: Windows venvs use a different structure, for unknown reasons.
+if [[ -d "${GITHUB_ACTION_PATH}/.action-env/bin" ]]; then
+  VENV_PYTHON_PATH="${GITHUB_ACTION_PATH}/.action-env/bin/python"
+else
+  VENV_PYTHON_PATH="${GITHUB_ACTION_PATH}/.action-env/Scripts/python"
+fi
+
+"${VENV_PYTHON_PATH}" -m pip install --requirement "${GITHUB_ACTION_PATH}/requirements.txt"
+
+debug "sigstore-python: $("${VENV_PYTHON_PATH}" -m sigstore --version)"
+
+# Finally, propagate VENV_PYTHON_PATH so we can actually kick-start
+# the extension from it.
+echo "venv-python-path=${VENV_PYTHON_PATH}" >> "${GITHUB_OUTPUT}"

--- a/setup/setup.bash
+++ b/setup/setup.bash
@@ -35,7 +35,7 @@ if [[ "${0}" == "${BASH_SOURCE[0]}" ]]; then
   die "Internal error: setup harness was executed instead of being sourced?"
 fi
 
-# Check the Python version, making sure it's new enough (3.7+)
+# Check the Python version, making sure it's new enough (3.8+)
 # The installation step immediately below will technically catch this,
 # but doing it explicitly gives us the opportunity to produce a better
 # error message.
@@ -43,7 +43,7 @@ vers=$(python -V | cut -d ' ' -f2)
 maj_vers=$(cut -d '.' -f1 <<< "${vers}")
 min_vers=$(cut -d '.' -f2 <<< "${vers}")
 
-[[ "${maj_vers}" == "3" && "${min_vers}" -ge 7 ]] || die "Bad Python version: ${vers}"
+[[ "${maj_vers}" == "3" && "${min_vers}" -ge 8 ]] || die "Bad Python version: ${vers}"
 
 # If the user didn't explicitly configure a Python version with
 # `actions/setup-python`, then we might be using the distribution's Python and

